src/RageSoundMixBuffer.cpp

```diff
@@ -31,19 +31,36 @@ void RageSoundMixBuffer::SetWriteOffset( int iOffset )
 	m_iOffset = iOffset;
 }
 
-void RageSoundMixBuffer::Extend( unsigned iSamples )
+void RageSoundMixBuffer::Extend(unsigned iSamples)
 {
-	const unsigned realsize = iSamples+m_iOffset;
-	if( m_iBufSize < realsize )
+	const std::uint64_t m_nBufSize = static_cast<std::uint64_t>(iSamples) + m_iOffset;
+	if (m_iBufSize < m_nBufSize)
 	{
-		m_pMixbuf = (float *) realloc( m_pMixbuf, sizeof(float) * realsize );
-		m_iBufSize = realsize;
+		m_pMixbufBackup = static_cast<float*>(realloc(m_pMixbufBackup, sizeof(float) * m_iBufSize));
```

Comment:
`realloc()` was used in the old code because the intention was to preserve the contents, but now there's no need to do that, and in fact a few lines below the contents get overwritten. Some other allocation function might be more appropriate here.

Reply:
I've tried quite a lot of forms of malloc, but they're not as robust as realloc here. It's quite difficult to manage backup buffers using malloc and ensure memory leaks won't happen. Other weird things will happen like the game will go silent half an hour into a set or something with malloc.

Unfortunately I've just squashed the history before you reviewed this, but I'd be happy to show some of the other things I tried here 🙂

Reply:
> I agree that `realloc()` does make memory management quite easy in this case. Since memory leaks are a concern, how about modernizing this part and using `std::unique_ptr`?
> 
> In any case, what `realloc()` does here, other than helping with memory management, is a copy of the old memory. And then just below we have a `memcpy()` which copies memory again. So I think that the alleged race condition in the code is still present, it's probably just hidden now by the delay caused by 2 additional memory copies.

As I mentioned here (https://github.com/itgmania/itgmania/pull/246#discussion_r1601592787) I don't believe that the issue is a race condition (though one may exist?), the issue is there is currently no method to recover if reallocation fails and leads to a null pointer. When it happens a very distracting (and preventable) stutter occurs if we are mid-game. Having the secondary buffer will prevent this distracting stutter.

This is very rare, but still happens once in a while and it will cause a stutter in game that it does eventually recover from. Having the secondary buffer helps it recover much faster and not interrupt gameplay when it happens.

Reply:
The original comment for the pull request mentioned the audio buffer can glitch if a `realloc()` is performed while the buffer is being read, which indicates a race condition. But if the actual issue is `realloc()` failing and returning `NULL`, then it would be good to check why that's happening. An immediate reason which comes to mind is being low on memory, but it's also possible that for some reason an unusually big buffer size is being requested. If it's the latter than this should probably be investigated.

Also as mentioned in https://github.com/itgmania/itgmania/pull/246#discussion_r1601627165 there is no need for a backup buffer in case `realloc()` fails.

Reply:
> I agree that `realloc()` does make memory management quite easy in this case. Since memory leaks are a concern, how about modernizing this part and using `std::unique_ptr`?

I apologize, I realized I never addressed this point specifically.

The issue with modernizing code this old is that the entirety of not only the SM5 code base but often this code hasn't been meaningfully changed since the early days of SM3 at best. There are multiple decades of code built on top of this stuff that all expects it behaves the way it does.

Unfortunately everything I tried in terms of replacing `realloc` / `malloc` / `memset` with more modern operations, such as using `std::unique_ptr` or `std::vector` caused the game to compile successfully but crash almost immediately upon arriving at the main screen. Stuff like this happened frequently during my work on #230 as well, in many cases with code this old things will break very strangely, if it even agrees to compile, when you start replacing functions with more modern ones.

For this reason I am very hesitant to do much more than update to C++ style casts until we are in a place where we are ready to do major replacements on a lot of this core code.